hail/python/hailtop/aiotools/validators.py

```diff
@@ -35,7 +35,7 @@ async def _async_validate_file(
     if isinstance(fs, GoogleStorageAsyncFS):
         location = fs.storage_location(uri)
         if location not in fs.allowed_storage_locations:
-            if not await fs.is_hot_storage(location):
+            if not await fs.is_hot_storage(location, uri):
```

Comment:
I think there's a subtle bug here for the following case:
```python
validate_file('gs://public-bucket/hot-storage-obj')  # OK because the object is hot storage
validate_file('gs://public-bucket/cold-storage-obj')  # doesn't error because public-bucket is cached in allowed_storage_locations
```

Whether we're querying the bucket or the object needs to somehow affect our caching strategy.

Reply:
Ah good point, I think it is OK to assume that storage classes are uniform across all objects in a bucket so long as we very explicitly document that. I don't know if there's a sound middle ground between that and check all the objects. I think the thing I most want to avoid is fetching metadata for all files in a 50k partition dataset.